tests/input.rs

```diff
@@ -0,0 +1,226 @@
+            scan_code: 0,
+            state: ButtonState::Released,
+        },
+    );
+
+    // Another button is clicked, the vec is empty
+    app.update();
+    assert!(inputs.lock().unwrap().is_empty());
+}
+
+#[test]
+fn test_input_multiple_received_char_and_keyboard_input_are_independent() {
+    let expected_received_chars = Arc::new(Mutex::new(VecDeque::from(['a', 'B'])));
+    let expected_keyboard_inputs = Arc::new(Mutex::new(VecDeque::from([
+        KeyboardInput {
+            key_code: Some(KeyCode::A),
+            scan_code: 0,
+            state: ButtonState::Pressed,
+        },
+        KeyboardInput {
+            key_code: Some(KeyCode::B),
+            scan_code: 0,
+            state: ButtonState::Released,
+        },
+    ])));
+    let expected_received_chars2 = expected_received_chars.clone();
+    let expected_keyboard_inputs2 = expected_keyboard_inputs.clone();
+    let keyboard_system = move |mut chat_input_events: EventReader<KeyboardInput>| {
+        for event in chat_input_events.iter() {
+            assert_eq!(
+                event.clone(),
+                expected_keyboard_inputs2
+                    .lock()
+                    .unwrap()
+                    .pop_front()
+                    .expect("Unexpected event")
+            );
+        }
+    };
+    let received_char_system = move |mut chat_input_events: EventReader<ReceivedCharacter>| {
+        for event in chat_input_events.iter() {
+            assert_eq!(
+                event.char,
+                expected_received_chars2
+                    .lock()
+                    .unwrap()
+                    .pop_front()
+                    .expect("Unexpected event")
+            );
+        }
+    };
+
+    let mut app = App::new();
+    app.add_plugin(InputPlugin::default())
+        .add_plugin(WindowPlugin::default())
+        .add_system(keyboard_system)
+        .add_system(received_char_system);
+
+    send_event(
+        &mut app,
+        KeyboardInput {
+            key_code: Some(KeyCode::A),
+            scan_code: 0,
+            state: ButtonState::Pressed,
+        },
+    );
+    send_event(
+        &mut app,
+        KeyboardInput {
+            key_code: Some(KeyCode::B),
+            scan_code: 0,
+            state: ButtonState::Released,
+        },
+    );
+    send_event(
+        &mut app,
+        ReceivedCharacter {
+            id: WindowId::primary(),
+            char: 'a',
+        },
+    );
+    send_event(
+        &mut app,
+        ReceivedCharacter {
+            id: WindowId::primary(),
+            char: 'B',
+        },
+    );
+
+    // Multiple events can be sent multiple time per `update`.
+    // The iterator inside the system function returns 2 events
+    app.update();
+    assert_eq!(expected_keyboard_inputs.lock().unwrap().len(), 0);
+    assert_eq!(expected_received_chars.lock().unwrap().len(), 0);
+}
+
+fn send_event<E: Event>(app: &mut App, event: E) {
```

Comment:
I think this should just be a method on `App` and `World`. Could you split this out into its own PR?